ray-on-gke/tpu/kuberay-tpu-webhook/main.go

```diff
@@ -357,39 +405,84 @@ func getReplicaIndex(clusterName string, groupName string, namespace string) int
 
 // returns next lowest TPU_WORKER_ID in pod slice and updates mappings
-func getNextWorkerID(podSlice slice, namespace string, replicaIndex int) int {
-	tpuWorkerID := 0
-	if sliceToWorkers[podSlice] == nil {
-		newWorker := worker{tpuWorkerID, replicaIndex, true}
-		sliceToWorkers[podSlice] = []worker{newWorker}
-	} else {
-		nextLowestID := math.MaxInt32
-		replacePod := false
-		// iterate through existing workers and check if any have been deleted
-		for _, worker := range sliceToWorkers[podSlice] {
-			if worker.isCreated == false && worker.workerIndex < nextLowestID {
-				replacePod = true
-				nextLowestID = worker.workerIndex
+func getNextWorkerID(sliceToWorkerIDs map[slice][]int, podSlice slice, namespace string, replicaIndex int) int {
+	tpuWorkerID := 0 // defaults to 0 (first Pod in slice)
+	if sliceToWorkerIDs[podSlice] != nil {
+		sort.Ints(sliceToWorkerIDs[podSlice])
+		// iterate through existing workers and get the next lowest, unused ID
+		for _, workerID := range sliceToWorkerIDs[podSlice] {
+			if workerID == tpuWorkerID {
+				tpuWorkerID++
 			}
 		}
-		// reassign next lowest TPU_WORKER_ID if pod has been deleted
-		if replacePod == true {
-			for index, worker := range sliceToWorkers[podSlice] {
-				// set worker.isCreated to true now that pod is being re-created
-				if worker.workerIndex == nextLowestID {
-					sliceToWorkers[podSlice][index].isCreated = true
+	}
+	klog.V(1).InfoS("getNextWorkerID", "RayCluster", namespace+"/"+podSlice.clusterName, "Worker Group", podSlice.groupName, "TPU_WORKER_ID", tpuWorkerID)
+	return tpuWorkerID
+}
+
+// builds mapping representing the current RayCluster state of TPU pods using PodInformer
+func getSliceToWorkerIDs(podLister listersv1.PodLister, clusterName string, groupName string, namespace string, numOfHosts int32) (map[slice][]int, error) {
+	sliceToWorkerIDs := make(map[slice][]int)
+
+	// retrieve list of Pods in the same Ray worker group as the intercepted Pod
+	if podLister == nil {
+		return nil, errors.New("k8s Pod Informer Lister not initialized")
+	}
+	podsInGroup, err := podLister.Pods(namespace).List(labels.SelectorFromSet(labels.Set{"ray.io/group": groupName}))
+	if err != nil {
+		return nil, err
+	}
+
+	if podsInGroup == nil {
+		// return an empty mapping if no Pods with 'ray.io/group' label found
+		return sliceToWorkerIDs, nil
+	}
+	klog.V(1).InfoS("getSliceToWorkerIDs", "RayCluster", namespace+"/"+clusterName, "# Pods in Group", len(podsInGroup))
+	for _, existingPod := range podsInGroup {
+		if existingPod.Status.Phase != "Pending" && existingPod.Status.Phase != "Running" {
+			continue
+		}
+		existingClusterName := existingPod.Labels["ray.io/cluster"]
+		existingGroupName := existingPod.Labels["ray.io/group"]
+		existingNamespace := existingPod.Namespace
+		// we only care about workers in the same RayCluster and worker group when assigning IDs
+		if clusterName == existingClusterName && groupName == existingGroupName && namespace == existingNamespace {
```

Comment:
This if and the if condition after with `containerRequestingTPUs` can be early returns

Reply:
Done in [dce1bbe](https://github.com/GoogleCloudPlatform/ai-on-gke/pull/723/commits/dce1bbe4b09fd0a48304cb2d519414578a3ecdc0), I changed the second if condition to an early `continue`, since if the Pod is missing the `replicaIndex` label that just means it hasn't been intercepted yet, and we still want to check the other Pods in the cache.